XIPs/xip-17-remote-attachment-content-type-proposal.md

```diff
@@ -0,0 +1,52 @@
+---
+xip: 17
+title: Remote Attachment Content Type
+description: Content type for remote attachments
+author: Pat Nakajima
+status: Review
+type: Standards Track
+category: XRC
+created: 2023-02-15
+---
+
+## Abstract
+
+This XRC proposes a new content type for messages that do not consist of text and/or are too large to be sent on the XMTP network. It builds on [XIP 15](https://github.com/xmtp/XIPs/pull/15)'s basic attachment type, but instead of messages being sent on the network, the message contains a URL. The contents of the URL is an `Envelope` that can be decoded by clients.
+
+## Motivation
+
+The current 1MB limit for messages on the XMTP network is too small for many types of content. To enable larger files, external storage is going to be necessary.
+
+## Specification
+
+Proposed content type:
+
+```js
+{
+  authorityId: "xmtp.org"
+  typeId: "remoteAttachment"
+  versionMajor: 0
+  versionMinor: 1
+}
+```
+
+The content of the encoded message is a URL that points to encoded Envelope data that would not be limited by the 1MB limit, since it's not being sent on the network. By using the existing `Envelope` format, we can still ensure E2E encryption for this data.
```

Comment:
Would be helpful to specify the encryption scheme. How do we pass keys to the reader? What types of keys are allowed?

Reply:
It would use the same encryption we use for publishing envelopes to the network, but instead of publishing to the network it would be to some other host.

Reply:
Which type, though. For V1 messages we use X3DH and for V2 messages we use random `TopicKey`s. In either case, we would need to figure out where to store the salt and the nonce. Those values are unique and random, and we wouldn't want to re-use the same values that went into the message since nonce re-use is an attack vector that could lead to messages being compromised.

So, we have a few options:
1. In the message payload, include the `nonce` and `salt` and use whatever secret was used to encrypt the message. The remote content would just be the encrypted bytes of the message
2. Keep the message payload light, and just include a link to the content. The remote content would need to be formatted as `xmtp.Ciphertext` proto message. Re-use the shared secret for the message (either V1 X3DH or V2 TopicKey)
3. Use an ephemeral key. The message payload would include the key (maybe 32 bytes of randomness), nonce, and salt. The remote message would be the encrypted payload only.

Of those options (those are just the three off the top of my head), I'd lean towards 3. Gives the most security, since each payload is encrypted with a different key, and means that we don't have to decode another protobuf to read the message.

Reply:
@jazzz @richardhuaaa @michaelx11 curious what the ProtoPod thinks

Reply:
I think maybe I'm missing something with how our encryption works. My thought for the remote payload is that it's an `xmtp.Envelope` with a topic, timestamp and message. In the same way we can send those as push notifications, I was thinking we could also make them the remote payload. It was my understanding that anyone can query for those but the only way to decrypt it is if you have the conversation key material as well (my thinking was that we'd be using the V2 message format).

Reply:
So, I guess we could do the full envelope. That's basically option 2 above. The topic and timestamp feel a little superfluous (won't they always be the same as the on-network message?), but are harmless.

Reply:
I like the full envelope because clients already need to be able to handle decoding those. I agree about the topic/timestamp being superfluous but I think it's worth it for the simplicity.

Reply:
I'd lean towards option 3. where:
- PointerMessage contains a symmetricKey and all the parameters to decrypt it.
- Server contains the content payload only.

I am weary of xmtp.Envelope as the abstraction interface. 
- All possible Envelopes stored remotely requires exponentially more testing. 
   - "What happens if a Key Revocation message is stored off network?",  
   - "what happens if an Invite is no longer accessible?"
   - "What if a remoteContent points to another removeContent?"
- Coupling these messages to a specific XMTP Message version would make future migrations more complicated.

Reply:
I hear you @jazzz on the potential complexity of having the remote payloads be some sort of Content Type. I'm unsold as `xmtp.Envelope` being the right wrapper. But an encrypted `EncodedContent` payload will make things a lot more usable for developers than a pile of unstructured bytes. Especially if these are messages that are meant to be displayed inline, where we would want to have the regular set of tools for decoding available that we offer for regular messages.

Otherwise you are going to have one flow to handle remote images, and a totally different decoding flow for on-network images. That seems less than ideal.

> "What happens if a Key Revocation message is stored off network?"

Since we don't have Key Revocation messages yet, hard for me to answer conclusively. Does seem like we may want certain meta-messages to live outside the normal message encoding flow and be typed based on contentTopic. We do this with `Invites` currently, which are not a `ContentType` at all.

> "what happens if an Invite is no longer accessible?"

Are you asking what happens if someone sends a conversation invite as a remote message? That should not be possible, as invites are separate from messages and are not `EncodedContent` or sent in message channels.

> "What if a remoteContent points to another removeContent?"

Both the spec and the example implementation need a mitigation for this kind of attack. Seems like we can have some sort of recursion limit in the SDK and throw an error after N loops.

Reply:
> But an encrypted EncodedContent payload will make things a lot more usable for developers than a pile of unstructured bytes.

I'll start by saying I can get behind this. EncodedContent payload seems like the an acceptable place an alevates many of my concerns.

-------------

To round out my thoughts: If the xmtp.Envelope wrapper is excluded then I see two options:

<img width="708" alt="remoteContent" src="https://user-images.githubusercontent.com/473256/220426788-bb056815-8040-472b-8c0a-d61ce6490d39.png">

*OptionA* -- Making remoteContent a composable type Leads to maximum utility. Any type can now be stored remotely with no changes required.

*OptionB* -- Making RemoteContent an implementation pattern for other type means it could be handled on a type by type basis.  Larger types (Images, Video, File) could support remoteContent but this is not required of all types.

My argument for B is not particularly strong: There are only few types that could really benefit from OptionA, yet potential remote storage may become a burden for future ContentType Developers . Any future ContentTypes will now need to be evaluated against a wider threat model, as independent ContentServers are outside of the protection of the Network consensus protocol. I don't have any hard examples as there are few formal XIP ContentTypes defined. Given this Option A is compelling.

Reply:
>We do this with Invites currently, which are not a ContentType at all.

Great point.

> That should not be possible

Great point. Despite the payload being able to hold an invite, it would not be possible to get an EncodedContent payload into the invite processing flow.

Reply:
I'm with @jazzz in that I'm hesitant to define a `RemoteContent` type that can wrap any normal `XMTP` content type. I worry about the possible attack surface caused by remote fetches such as SSRF (e.g. attacker puts 127.0.0.1:[some important port]/[some sensitive path]). We'd have to consider the cross product of `All RemoteContent actions X All ContentTypes`. I'd be more comfortable if we keep `RemoteContent`'s capabilities very well described and limited. For example: must be a GET over HTTPS.